stored_requests/backends/file_fetcher/fetcher.go

```diff
@@ -15,30 +20,80 @@ import (
 //
 // This expects each file in the directory to be named "{config_id}.json".
 // For example, when asked to fetch the request with ID == "23", it will return the data from "directory/23.json".
-func NewFileFetcher(directory string) (stored_requests.CacheableFetcher, error) {
+//
+// Providing a value of true for `watch` will monitor loaded files for changes and reload as
+// necessary.
+func NewFileFetcher(directory string, watch bool) (stored_requests.CacheableFetcher, error) {
 	fileInfos, err := ioutil.ReadDir(directory)
 	if err != nil {
 		return nil, err
 	}
-	storedReqs := make(map[string]json.RawMessage, len(fileInfos))
-	for _, fileInfo := range fileInfos {
-		fileData, err := ioutil.ReadFile(fmt.Sprintf("%s/%s", directory, fileInfo.Name()))
+
+	var watcher *fsnotify.Watcher
+	if watch {
+		watcher, err = fsnotify.NewWatcher()
 		if err != nil {
 			return nil, err
 		}
-		if strings.HasSuffix(fileInfo.Name(), ".json") { // Skip the .gitignore
+	}
+
+	storedReqs := make(map[string]json.RawMessage, len(fileInfos))
+	for _, fileInfo := range fileInfos {
+		filePath := path.Join(directory, fileInfo.Name())
+		if strings.HasSuffix(fileInfo.Name(), ".json") {
+			fileData, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				return nil, err
+			}
 			storedReqs[strings.TrimSuffix(fileInfo.Name(), ".json")] = json.RawMessage(fileData)
+			if watch {
+				if err = watcher.Add(filePath); err != nil {
+					return nil, err
+				}
+			}
 		}
 	}
-	return &eagerFetcher{
+
+	fetcher := &eagerFetcher{
 		stored_requests.Subscriptions{},
 		storedReqs,
-	}, nil
+		watcher,
+	}
+
+	if watch {
+		go func() {
+			for {
+				select {
+				case event := <-watcher.Events:
+					if event.Op&fsnotify.Write == fsnotify.Write {
+						filePath := event.Name
+						glog.Infof("Reloading file: %s", filePath)
+						fileData, err := ioutil.ReadFile(filePath)
+						if err != nil {
+							glog.Errorf("Error reloading file: %v", err)
+						}
+						fileName := path.Base(filePath)
+						id := strings.TrimSuffix(fileName, ".json")
+						update := map[string]json.RawMessage{id: fileData}
+						fetcher.storedReqs[id] = update[id]
```

Comment:
This should be safe, I don't believe any writes could happen concurrently as this is the single goroutine handling these fsnotify events for all files being watched.